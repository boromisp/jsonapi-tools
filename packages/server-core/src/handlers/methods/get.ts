'use strict';

import * as bluebird from 'bluebird';

import getRelatedSchema from './internal/get-related-schema';
import modelForType from './internal/model-for-type';
import dataToLinkage from './internal/data-to-linkage';
import dataToResource from './internal/data-to-resource';

import CustomError from '../../utils/custom-error';

import { IModel, IModels } from '../../types/model';

import { IRequestParamsBase } from './types/request-params';

import {
  IResourceObject,
  IRelationshipObject,
  ISuccessResourceDocument,
  ISuccessRelationshipDocument,
  IResourceIdentifierObject
} from 'jsonapi-types';
import { IParsedQueryFields, IParsedIncludes, ISuccessResponseObject } from '../../types/utils';

type IGetRest = Pick<IGetRequestParamsBase, 'method' | 'page' | 'options'>;

function getResource(
  model: IModel,
  id: string,
  fields: IParsedQueryFields | null,
  rest: IGetRest
): PromiseLike<IResourceObject | null> {
  return model.getOne(Object.assign({
    id,
    fields: fields && fields[model.schema.type]
  }, rest)).then(data => dataToResource(model.schema, data));
}

function getResources(
  model: IModel,
  ids: string[] | null,
  fields: IParsedQueryFields | null,
  filters: object | null,
  sorts: string[] | null,
  rest: IGetRest
): PromiseLike<IResourceObject[]> {
  if (ids === null) {
    return model.getAll(Object.assign({
      fields: fields && fields[model.schema.type],
      filters,
      sorts
    }, rest)).then(data => data ? data.map(item => dataToResource(model.schema, item)) : []);
  } else {
    return model.getSome(Object.assign({
      ids,
      fields: fields && fields[model.schema.type],
      filters,
      sorts
    }, rest)).then(data => data ? data.map(item => dataToResource(model.schema, item)) : []);
  }
}

function getRelationshipObject(
  model: IModel,
  id: string,
  relationship: string,
  fields: IParsedQueryFields | null,
  rest: IGetRest): PromiseLike<IRelationshipObject> {

  return bluebird.try(() => {
    const relatedSchema = getRelatedSchema(model.schema, relationship);
    const type = model.schema.type;
    const localFields = fields && fields[type] || new Set();
    localFields.add(relationship);

    return model.getOne(Object.assign({ id, fields: localFields }, rest)).then(data => {
      if (data) {
        return dataToLinkage(relatedSchema, data[relationship], type, id, relationship);
      }
      throw new CustomError(`Parent object for relationship ${type}(${id}).${relationship} not found.`, 404);
    });
  });
}

function validateIncludes(models: IModels, type: string, includes: IParsedIncludes) {
  const relationships = models[type].schema.relationships || {};
  for (const relationship of Object.keys(includes)) {
    if (!relationships[relationship]) {
      throw new CustomError(`Invalid include. Type ${type} has no relationship named ${relationship}.`, 400);
    }
    validateIncludes(models, relationships[relationship].type, includes[relationship]);
  }
}

type TypeLinkMap = Map<string, Set<string>>;

function buildLinkMap(links: Array<IResourceObject | IResourceIdentifierObject>, linkMap: TypeLinkMap = new Map()) {
  for (const link of links) {
    if (!linkMap.has(link.type)) {
      linkMap.set(link.type, new Set());
    }
    linkMap.get(link.type)!.add(link.id);
  }
  return linkMap;
}

function uniqueLinks(links: Array<IResourceObject | IResourceIdentifierObject>): IResourceIdentifierObject[] {
  const out: Array<{ type: string, id: string }> = [];
  for (const [type, ids] of buildLinkMap(links)) {
    for (const id of ids) {
      out.push({ type, id });
    }
  }
  return out;
}

type TypeItemCache = Map<string, Map<string, IResourceObject>>;

function addToItemCache(itemCache: TypeItemCache, resource: IResourceObject) {
  if (!itemCache.has(resource.type)) {
    itemCache.set(resource.type, new Map());
  }
  itemCache.get(resource.type)!.set(resource.id, resource);
}

function includeTier(
  models: IModels,
  fields: IParsedQueryFields | null,
  tierNodes: IParsedIncludes[],
  itemCache: TypeItemCache,
  includeNodeLinks: Map<IParsedIncludes, IResourceObject[]>,
  rest: IGetRest): PromiseLike<void> {
  return bluebird.try(() => {
    const tierLinks: TypeLinkMap = new Map();

    for (const node of tierNodes) {
      // TODO: includeNodeLinks.get(node) what if not found?
      const nodeParents = includeNodeLinks.get(node)!.map(identifier => {
        const resourcesWithType = itemCache.get(identifier.type);
        if (resourcesWithType) {
          const parent = resourcesWithType.get(identifier.id);
          if (parent) {
            return parent;
          }
        }
        throw new Error('Parent node missing with type: ' + identifier.type + ' and id: ' + identifier.id + '.');
      }).filter(Boolean);

      for (const relationship of Object.keys(node)) {
        const nodeLinks = uniqueLinks(nodeParents.reduce((links: IResourceIdentifierObject[], resource) => {
          if (resource.relationships && resource.relationships[relationship]) {
            const relationshipData = resource.relationships[relationship].data;
            if (relationshipData) {
              if (Array.isArray(relationshipData)) {
                return links.concat(relationshipData);
              }
              links.push(relationshipData);
            }
          }
          return links;
        }, []));
        includeNodeLinks.set(node[relationship], nodeLinks);
        buildLinkMap(nodeLinks, tierLinks);
      }
    }

    const newLinks = [...tierLinks].reduce((links: Array<[IModel, string[]]>, [type, ids]) => {
      const newIDs = [...ids].filter(id => !itemCache.has(type) || !itemCache.get(type)!.has(id));
      if (newIDs.length > 0) {
        links.push([modelForType(models, type), newIDs]);
      }
      return links;
    }, []);

    const promises = newLinks.map(([model, ids]) => getResources(model, ids, fields, null, null, rest));
    return bluebird.all(promises);
  }).then(resourceArrays => {
    for (const resources of resourceArrays) {
      for (const resource of resources) {
        addToItemCache(itemCache, resource);
      }
    }

    const nextTierNodes = tierNodes.reduce((nodes: IParsedIncludes[], node) => {
      for (const relationship of Object.keys(node)) {
        if (Object.keys(node[relationship]).length) {
          nodes.push(node[relationship]);
        }
      }
      return nodes;
    }, []);

    if (nextTierNodes.length === 0) {
      return;
    }

    return includeTier(models, fields, nextTierNodes, itemCache, includeNodeLinks, rest);
  });
}

function getIncludedResources(
  models: IModels,
  primary: IResourceObject[],
  fields: IParsedQueryFields | null,
  includes: IParsedIncludes,
  rest: IGetRest
): PromiseLike<IResourceObject[]> {
  return bluebird.try(() => {
    const primaryType = primary[0].type;

    validateIncludes(models, primaryType, includes);

    const itemCache: TypeItemCache = new Map();
    for (const resource of primary) {
      addToItemCache(itemCache, resource);
    }

    const includeNodeLinks: Map<IParsedIncludes, IResourceObject[]> = new Map();
    includeNodeLinks.set(includes, primary);

    return includeTier(models, fields, [includes], itemCache, includeNodeLinks, rest).then(() => {
      const included: IResourceObject[] = [];
      // TODO: what if includeNodeLinks doesn't contain includes?
      const primaryLinks = buildLinkMap(includeNodeLinks.get(includes)!);

      for (const [type, ids] of itemCache) {
        for (const [id, resource] of ids) {
          if (!primaryLinks.has(type) || !primaryLinks.get(type)!.has(id)) {
            included.push(resource);
          }
        }
      }

      return included;
    });
  });
}

export interface IGetRequestParamsBase extends IRequestParamsBase {
  method: 'get';
  page: object | null;
  type: string;
  models: IModels;
}

export interface IGetOneRequestParams extends IGetRequestParamsBase {
  id: string;
  fields: IParsedQueryFields | null;
  includes: IParsedIncludes | null;
}

export interface IGetAllRequestParams extends IGetRequestParamsBase {
  filters: object | null;
  sorts: string[] | null;
  fields: IParsedQueryFields | null;
  includes: IParsedIncludes | null;
}

export interface IGetRelationshipRequestParams extends IGetRequestParamsBase {
  id: string;
  relationship: string;
  asLink: true;
}

export interface IGetRelatedResourceRequestParams extends IGetAllRequestParams {
  id: string;
  relationship: string;
  asLink: false;
}

export type IGetRequestParams = IGetOneRequestParams
  | IGetAllRequestParams
  | IGetRelationshipRequestParams
  | IGetRelatedResourceRequestParams;

function isGetOneRequest(request: IGetRequestParams): request is IGetOneRequestParams {
  return (request as IGetOneRequestParams).id !== undefined;
}

function isRelatedRequest(
  request: IGetRequestParams
): request is IGetRelationshipRequestParams | IGetRelatedResourceRequestParams {
  return isGetOneRequest(request)
    && (request as IGetRelationshipRequestParams | IGetRelatedResourceRequestParams).relationship !== undefined;
}

function includeRelatedResources(
  requestParams: IGetOneRequestParams | IGetAllRequestParams | IGetRelatedResourceRequestParams,
  top: ISuccessResourceDocument
): PromiseLike<ISuccessResourceDocument> {
  const { models, fields, includes, options, page, method } = requestParams;
  const rest = { options, page, method};

  if (includes && top.data && (!Array.isArray(top.data) || top.data.length)) {
    const primary = Array.isArray(top.data) ? top.data : [top.data];
    return getIncludedResources(models, primary, fields, includes, rest).then(included => {
      if (included.length) {
        top.included = included;
      }
      return top;
    });
  }
  return bluebird.resolve(top);
}

function getRelatedResourceDocument(
  requestParams: IGetRelatedResourceRequestParams
): PromiseLike<ISuccessResourceDocument> {
  const { models, type, fields, options, page, method, filters, sorts, id, relationship } = requestParams;
  const rest = { options, page, method};

  return bluebird.try(() => modelForType(models, type))
    .then(model => getRelationshipObject(model, id, relationship, null, rest))
    .then(relationshipObject => {
      const top: ISuccessResourceDocument = { data: null };
      const relatedLink = relationshipObject.links && relationshipObject.links.related;
      if (relatedLink) {
        top.links = { self: relatedLink };
      }
      if (Array.isArray(relationshipObject.data)) {
        if (relationshipObject.data.length) {
          return getResources(
            modelForType(models, relationshipObject.data[0].type),
            relationshipObject.data.map(item => item.id),
            fields, filters, sorts,
            rest
          ).then(data => { top.data = data; return top; });
        }
        top.data = [];
      } else if (relationshipObject.data) {
        return getResource(
          modelForType(models, relationshipObject.data.type),
          relationshipObject.data.id,
          fields,
          rest
        ).then(data => { top.data = data; return top; });
      }
      return top;
    }).then(top => includeRelatedResources(requestParams, top));
}

function getRelationshipDocument(
  requestParams: IGetRelationshipRequestParams
): PromiseLike<ISuccessRelationshipDocument> {
  const { models, type, options, page, method, id, relationship } = requestParams;
  const rest = { options, page, method};

  return bluebird.try(() => modelForType(models, type))
    .then(model => getRelationshipObject(model, id, relationship, null, rest));
}

function getResourceDocument(requestParams: IGetOneRequestParams): PromiseLike<ISuccessResourceDocument> {
  const { models, type, fields, options, page, method } = requestParams;
  const rest = { options, page, method};

  return bluebird.try(() => modelForType(models, type))
    .then(model => getResource(model, requestParams.id, fields, rest))
    .then(resource => {
      if (!resource) {
        throw new CustomError(`Item of type ${type} with id ${requestParams.id} not found.`, 404);
      }
      const top: ISuccessResourceDocument = { data: resource };
      if (resource.links) {
        top.links = { self: resource.links.self };
      }
      return top;
    });
}

function getResourcesDocument(requestParams: IGetAllRequestParams): PromiseLike<ISuccessResourceDocument> {
  const { models, type, fields, options, page, method } = requestParams;
  const rest = { options, page, method};

  return bluebird.try(() => modelForType(models, type))
    .then(model => getResources(model, null, fields, requestParams.filters, requestParams.sorts, rest)
      .then(resources => {
         return {
          links: model.schema.links ? model.schema.links({ id: false }) : {
            self: `/${type}`
          },
          data: resources
        } as ISuccessResourceDocument;
      }));
}

function getResponseDocument(
  requestParams: IGetRequestParams
): PromiseLike<ISuccessResourceDocument | ISuccessRelationshipDocument> {
  if (isRelatedRequest(requestParams)) {
    return requestParams.asLink
      ? getRelationshipDocument(requestParams)
      : getRelatedResourceDocument(requestParams);
  }
  return isGetOneRequest(requestParams)
    ? getResourceDocument(requestParams)
    : getResourcesDocument(requestParams);
}

export default function get(requestParams: IGetRequestParams): PromiseLike<ISuccessResponseObject> {
  return getResponseDocument(requestParams).then(top => ({ status: 200, body: top }));
}
