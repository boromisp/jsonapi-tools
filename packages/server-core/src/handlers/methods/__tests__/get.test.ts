import 'jest';

import get, { IGetOneRequestParams, IGetAllRequestParams } from '../get';
import { ISuccessResponseObject } from '../../../types/utils';
import { IModel, IModels, ISchema } from '../../../types/model';

import serializers from '../../../__test_utils__/custom-serializers';

for (const serializer of serializers) {
  expect.addSnapshotSerializer(serializer);
}

function mockModel(schema: IModel['schema'] | string, empty?: boolean): IModel {
  return {
    schema: typeof schema === 'string' ? { type: schema } : schema,
    getOne({ id }) {
      return Promise.resolve(empty ? null : { id });
    },
    getSome({ ids }) {
      return empty ? Promise.resolve([]) : Promise.all(ids.map(id => this.getOne({ id })));
    },
    getAll() {
      return this.getSome({ ids: ['1', '2', '3', '4', '5'] });
    },
    update() { return Promise.resolve(null); },
    create() { return Promise.resolve({ id: '1' }); },
    delete() { return Promise.resolve(true); }
  };
}

function buildModels(...args: IModel[]): IModels {
  return args.reduce((models, model) => {
    models[model.schema.type] = model;
    return models;
  }, {} as IModels);
}

describe('Test getting a single resource', () => {
  let requestParams: IGetOneRequestParams;

  beforeEach(() => {
    requestParams = {
      method: 'get',
      type: 'my-type',
      id: '1',
      fields: null,
      includes: null,
      page: null,
      options: {},
      models: {}
    };
  });

  test('Throws if model not found', () => {
    expect.assertions(1);
    return get(requestParams).then(null,
      error => expect(error).toMatchSnapshot());
  });

  test('Throws 404 if null returned', () => {
    requestParams.models = buildModels(mockModel('my-type', true));

    expect.assertions(1);
    return get(requestParams).then(null,
      error => expect(error).toMatchSnapshot());
  });

  test('Creates response document on success', () => {
    requestParams.models = buildModels(mockModel('my-type'));

    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Throws on invalid include', () => {
    const model1 = mockModel({
      type: 'my-type',
      relationships: {
        rel1: {
          type: 'other-type'
        }
      }
    });
    requestParams.models = buildModels(model1, mockModel('other-type'));
    requestParams.includes = { rel1: { rel2: {} } };
    expect.assertions(1);
    return get(requestParams).then(null, error => expect(error).toMatchSnapshot());
  });

  test('Includes related resource', () => {
    const model1 = mockModel({
      type: 'my-type',
      relationships: {
        rel1: {
          type: 'other-type'
        }
      }
    });
    model1.getOne = ({ id }) => Promise.resolve({ id, rel1: '1' });
    requestParams.models = buildModels(model1, mockModel('other-type'));
    requestParams.includes = { rel1: {} };
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Includes related resources', () => {
    const model1 = mockModel({
      type: 'my-type',
      relationships: {
        rel1: {
          type: 'other-type'
        }
      }
    });
    model1.getOne = ({ id }) => Promise.resolve({ id, rel1: ['1', '2', '3'] });
    requestParams.models = buildModels(model1, mockModel('other-type'));
    requestParams.includes = { rel1: {} };
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Sets null relationship data', () => {
    const model1 = mockModel({
      type: 'my-type',
      relationships: {
        rel1: {
          type: 'other-type'
        }
      }
    });
    model1.getOne = ({ id }) => Promise.resolve({ id, rel1: null });
    requestParams.models = buildModels(model1, mockModel('other-type'));
    requestParams.includes = { rel1: {} };
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Includes attributes', () => {
    const model = mockModel('my-type');
    model.getOne = ({ id }) => Promise.resolve({
      id,
      attr: '1',
      field: ['2', '3'],
      foo: { bar: new Date(1234567890123) }
    });
    requestParams.models = buildModels(model);
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });
});

describe('Test getting all resources', () => {
  let requestParams: IGetAllRequestParams;

  beforeEach(() => {
    requestParams = {
      method: 'get',
      type: 'my-type',
      fields: null,
      includes: null,
      page: null,
      filters: null,
      sorts: null,
      options: {},
      models: {}
    };
  });

  test('Throws if model not found', () => {
    expect.assertions(1);
    return get(requestParams).then(null,
      error => expect(error).toMatchSnapshot());
  });

  test('Returns body with empty data array if no items found', () => {
    requestParams.models = buildModels(mockModel('my-type', true));

    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Returns body with all items found', () => {
    requestParams.models = buildModels(mockModel('my-type'));

    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Includes related resource', () => {
    const model1 = mockModel({
      type: 'my-type',
      relationships: {
        rel1: {
          type: 'other-type'
        }
      }
    });
    model1.getOne = ({ id }) => Promise.resolve({ id, rel1: String(100 - Number(id)) });
    requestParams.models = buildModels(model1, mockModel('other-type'));
    requestParams.includes = { rel1: {} };
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Includes related resources', () => {
    const model1 = mockModel({
      type: 'my-type',
      relationships: {
        rel1: {
          type: 'other-type'
        }
      }
    });
    model1.getOne = ({ id }) => Promise.resolve({ id, rel1: [
      String(100 - Number(id) * 5),
      String(100 - Number(id) * 5),
      String(100 - Number(id) * 5)
    ] });
    requestParams.models = buildModels(model1, mockModel('other-type'));
    requestParams.includes = { rel1: {} };
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Sets null relationship data', () => {
    const model1 = mockModel({
      type: 'my-type',
      relationships: {
        rel1: {
          type: 'other-type'
        }
      }
    });
    model1.getOne = ({ id }) => Promise.resolve({ id, rel1: null });
    requestParams.models = buildModels(model1, mockModel('other-type'));
    requestParams.includes = { rel1: {} };
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });

  test('Includes attributes', () => {
    const model = mockModel('my-type');
    model.getOne = ({ id }) => Promise.resolve({
      id,
      attr: '1',
      field: ['2', '3'],
      foo: { bar: new Date(1234567890123) }
    });
    requestParams.models = buildModels(model);
    expect.assertions(1);
    return get(requestParams).then(doc => expect(doc).toMatchSnapshot());
  });
});
