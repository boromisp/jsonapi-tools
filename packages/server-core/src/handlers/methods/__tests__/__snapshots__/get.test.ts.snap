// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test getting a single resource Creates response document on success 1`] = `
Object {
  "body": Object {
    "data": Object {
      "id": "1",
      "links": Object {
        "self": "/my-type/1",
      },
      "type": "my-type",
    },
    "links": Object {
      "self": "/my-type/1",
    },
  },
  "status": 200,
}
`;

exports[`Test getting a single resource Includes attributes 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "attr": "1",
        "field": Array [
          "2",
          "3",
        ],
        "foo": Object {
          "bar": 2009-02-13T23:31:30.123Z,
        },
      },
      "id": "1",
      "links": Object {
        "self": "/my-type/1",
      },
      "type": "my-type",
    },
    "links": Object {
      "self": "/my-type/1",
    },
  },
  "status": 200,
}
`;

exports[`Test getting a single resource Includes related resource 1`] = `
Object {
  "body": Object {
    "data": Object {
      "id": "1",
      "links": Object {
        "self": "/my-type/1",
      },
      "relationships": Object {
        "rel1": Object {
          "data": Object {
            "id": "1",
            "type": "other-type",
          },
          "links": Object {
            "related": "/my-type/1/rel1",
            "self": "/my-type/1/relationships/rel1",
          },
        },
      },
      "type": "my-type",
    },
    "included": Array [
      Object {
        "id": "1",
        "links": Object {
          "self": "/other-type/1",
        },
        "type": "other-type",
      },
    ],
    "links": Object {
      "self": "/my-type/1",
    },
  },
  "status": 200,
}
`;

exports[`Test getting a single resource Includes related resources 1`] = `
Object {
  "body": Object {
    "data": Object {
      "id": "1",
      "links": Object {
        "self": "/my-type/1",
      },
      "relationships": Object {
        "rel1": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "other-type",
            },
            Object {
              "id": "2",
              "type": "other-type",
            },
            Object {
              "id": "3",
              "type": "other-type",
            },
          ],
          "links": Object {
            "related": "/my-type/1/rel1",
            "self": "/my-type/1/relationships/rel1",
          },
        },
      },
      "type": "my-type",
    },
    "included": Array [
      Object {
        "id": "1",
        "links": Object {
          "self": "/other-type/1",
        },
        "type": "other-type",
      },
      Object {
        "id": "2",
        "links": Object {
          "self": "/other-type/2",
        },
        "type": "other-type",
      },
      Object {
        "id": "3",
        "links": Object {
          "self": "/other-type/3",
        },
        "type": "other-type",
      },
    ],
    "links": Object {
      "self": "/my-type/1",
    },
  },
  "status": 200,
}
`;

exports[`Test getting a single resource Sets null relationship data 1`] = `
Object {
  "body": Object {
    "data": Object {
      "id": "1",
      "links": Object {
        "self": "/my-type/1",
      },
      "relationships": Object {
        "rel1": Object {
          "data": null,
          "links": Object {
            "related": "/my-type/1/rel1",
            "self": "/my-type/1/relationships/rel1",
          },
        },
      },
      "type": "my-type",
    },
    "links": Object {
      "self": "/my-type/1",
    },
  },
  "status": 200,
}
`;

exports[`Test getting a single resource Throws 404 if null returned 1`] = `[CustomError: Item of type my-type with id 1 not found., status: 404]`;

exports[`Test getting a single resource Throws if model not found 1`] = `[CustomError: Model for type my-type could not be found., status: 404]`;

exports[`Test getting a single resource Throws on invalid include 1`] = `[CustomError: Invalid include. Type other-type has no relationship named rel2., status: 400]`;

exports[`Test getting all resources Includes attributes 1`] = `
Object {
  "body": Object {
    "data": Array [
      Object {
        "attributes": Object {
          "attr": "1",
          "field": Array [
            "2",
            "3",
          ],
          "foo": Object {
            "bar": 2009-02-13T23:31:30.123Z,
          },
        },
        "id": "1",
        "links": Object {
          "self": "/my-type/1",
        },
        "type": "my-type",
      },
      Object {
        "attributes": Object {
          "attr": "1",
          "field": Array [
            "2",
            "3",
          ],
          "foo": Object {
            "bar": 2009-02-13T23:31:30.123Z,
          },
        },
        "id": "2",
        "links": Object {
          "self": "/my-type/2",
        },
        "type": "my-type",
      },
      Object {
        "attributes": Object {
          "attr": "1",
          "field": Array [
            "2",
            "3",
          ],
          "foo": Object {
            "bar": 2009-02-13T23:31:30.123Z,
          },
        },
        "id": "3",
        "links": Object {
          "self": "/my-type/3",
        },
        "type": "my-type",
      },
      Object {
        "attributes": Object {
          "attr": "1",
          "field": Array [
            "2",
            "3",
          ],
          "foo": Object {
            "bar": 2009-02-13T23:31:30.123Z,
          },
        },
        "id": "4",
        "links": Object {
          "self": "/my-type/4",
        },
        "type": "my-type",
      },
      Object {
        "attributes": Object {
          "attr": "1",
          "field": Array [
            "2",
            "3",
          ],
          "foo": Object {
            "bar": 2009-02-13T23:31:30.123Z,
          },
        },
        "id": "5",
        "links": Object {
          "self": "/my-type/5",
        },
        "type": "my-type",
      },
    ],
    "links": Object {
      "self": "/my-type",
    },
  },
  "status": 200,
}
`;

exports[`Test getting all resources Includes related resource 1`] = `
Object {
  "body": Object {
    "data": Array [
      Object {
        "id": "1",
        "links": Object {
          "self": "/my-type/1",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Object {
              "id": "99",
              "type": "other-type",
            },
            "links": Object {
              "related": "/my-type/1/rel1",
              "self": "/my-type/1/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "2",
        "links": Object {
          "self": "/my-type/2",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Object {
              "id": "98",
              "type": "other-type",
            },
            "links": Object {
              "related": "/my-type/2/rel1",
              "self": "/my-type/2/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "3",
        "links": Object {
          "self": "/my-type/3",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Object {
              "id": "97",
              "type": "other-type",
            },
            "links": Object {
              "related": "/my-type/3/rel1",
              "self": "/my-type/3/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "4",
        "links": Object {
          "self": "/my-type/4",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Object {
              "id": "96",
              "type": "other-type",
            },
            "links": Object {
              "related": "/my-type/4/rel1",
              "self": "/my-type/4/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "5",
        "links": Object {
          "self": "/my-type/5",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Object {
              "id": "95",
              "type": "other-type",
            },
            "links": Object {
              "related": "/my-type/5/rel1",
              "self": "/my-type/5/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
    ],
    "included": Array [
      Object {
        "id": "99",
        "links": Object {
          "self": "/other-type/99",
        },
        "type": "other-type",
      },
      Object {
        "id": "98",
        "links": Object {
          "self": "/other-type/98",
        },
        "type": "other-type",
      },
      Object {
        "id": "97",
        "links": Object {
          "self": "/other-type/97",
        },
        "type": "other-type",
      },
      Object {
        "id": "96",
        "links": Object {
          "self": "/other-type/96",
        },
        "type": "other-type",
      },
      Object {
        "id": "95",
        "links": Object {
          "self": "/other-type/95",
        },
        "type": "other-type",
      },
    ],
    "links": Object {
      "self": "/my-type",
    },
  },
  "status": 200,
}
`;

exports[`Test getting all resources Includes related resources 1`] = `
Object {
  "body": Object {
    "data": Array [
      Object {
        "id": "1",
        "links": Object {
          "self": "/my-type/1",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Array [
              Object {
                "id": "95",
                "type": "other-type",
              },
              Object {
                "id": "95",
                "type": "other-type",
              },
              Object {
                "id": "95",
                "type": "other-type",
              },
            ],
            "links": Object {
              "related": "/my-type/1/rel1",
              "self": "/my-type/1/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "2",
        "links": Object {
          "self": "/my-type/2",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Array [
              Object {
                "id": "90",
                "type": "other-type",
              },
              Object {
                "id": "90",
                "type": "other-type",
              },
              Object {
                "id": "90",
                "type": "other-type",
              },
            ],
            "links": Object {
              "related": "/my-type/2/rel1",
              "self": "/my-type/2/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "3",
        "links": Object {
          "self": "/my-type/3",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Array [
              Object {
                "id": "85",
                "type": "other-type",
              },
              Object {
                "id": "85",
                "type": "other-type",
              },
              Object {
                "id": "85",
                "type": "other-type",
              },
            ],
            "links": Object {
              "related": "/my-type/3/rel1",
              "self": "/my-type/3/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "4",
        "links": Object {
          "self": "/my-type/4",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Array [
              Object {
                "id": "80",
                "type": "other-type",
              },
              Object {
                "id": "80",
                "type": "other-type",
              },
              Object {
                "id": "80",
                "type": "other-type",
              },
            ],
            "links": Object {
              "related": "/my-type/4/rel1",
              "self": "/my-type/4/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "5",
        "links": Object {
          "self": "/my-type/5",
        },
        "relationships": Object {
          "rel1": Object {
            "data": Array [
              Object {
                "id": "75",
                "type": "other-type",
              },
              Object {
                "id": "75",
                "type": "other-type",
              },
              Object {
                "id": "75",
                "type": "other-type",
              },
            ],
            "links": Object {
              "related": "/my-type/5/rel1",
              "self": "/my-type/5/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
    ],
    "included": Array [
      Object {
        "id": "95",
        "links": Object {
          "self": "/other-type/95",
        },
        "type": "other-type",
      },
      Object {
        "id": "90",
        "links": Object {
          "self": "/other-type/90",
        },
        "type": "other-type",
      },
      Object {
        "id": "85",
        "links": Object {
          "self": "/other-type/85",
        },
        "type": "other-type",
      },
      Object {
        "id": "80",
        "links": Object {
          "self": "/other-type/80",
        },
        "type": "other-type",
      },
      Object {
        "id": "75",
        "links": Object {
          "self": "/other-type/75",
        },
        "type": "other-type",
      },
    ],
    "links": Object {
      "self": "/my-type",
    },
  },
  "status": 200,
}
`;

exports[`Test getting all resources Returns body with all items found 1`] = `
Object {
  "body": Object {
    "data": Array [
      Object {
        "id": "1",
        "links": Object {
          "self": "/my-type/1",
        },
        "type": "my-type",
      },
      Object {
        "id": "2",
        "links": Object {
          "self": "/my-type/2",
        },
        "type": "my-type",
      },
      Object {
        "id": "3",
        "links": Object {
          "self": "/my-type/3",
        },
        "type": "my-type",
      },
      Object {
        "id": "4",
        "links": Object {
          "self": "/my-type/4",
        },
        "type": "my-type",
      },
      Object {
        "id": "5",
        "links": Object {
          "self": "/my-type/5",
        },
        "type": "my-type",
      },
    ],
    "links": Object {
      "self": "/my-type",
    },
  },
  "status": 200,
}
`;

exports[`Test getting all resources Returns body with empty data array if no items found 1`] = `
Object {
  "body": Object {
    "data": Array [],
    "links": Object {
      "self": "/my-type",
    },
  },
  "status": 200,
}
`;

exports[`Test getting all resources Sets null relationship data 1`] = `
Object {
  "body": Object {
    "data": Array [
      Object {
        "id": "1",
        "links": Object {
          "self": "/my-type/1",
        },
        "relationships": Object {
          "rel1": Object {
            "data": null,
            "links": Object {
              "related": "/my-type/1/rel1",
              "self": "/my-type/1/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "2",
        "links": Object {
          "self": "/my-type/2",
        },
        "relationships": Object {
          "rel1": Object {
            "data": null,
            "links": Object {
              "related": "/my-type/2/rel1",
              "self": "/my-type/2/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "3",
        "links": Object {
          "self": "/my-type/3",
        },
        "relationships": Object {
          "rel1": Object {
            "data": null,
            "links": Object {
              "related": "/my-type/3/rel1",
              "self": "/my-type/3/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "4",
        "links": Object {
          "self": "/my-type/4",
        },
        "relationships": Object {
          "rel1": Object {
            "data": null,
            "links": Object {
              "related": "/my-type/4/rel1",
              "self": "/my-type/4/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
      Object {
        "id": "5",
        "links": Object {
          "self": "/my-type/5",
        },
        "relationships": Object {
          "rel1": Object {
            "data": null,
            "links": Object {
              "related": "/my-type/5/rel1",
              "self": "/my-type/5/relationships/rel1",
            },
          },
        },
        "type": "my-type",
      },
    ],
    "links": Object {
      "self": "/my-type",
    },
  },
  "status": 200,
}
`;

exports[`Test getting all resources Throws if model not found 1`] = `[CustomError: Model for type my-type could not be found., status: 404]`;
